File: scripts/generate_data.py
"""
FORE — scripts/generate_data.py
TASK-004 (Vishvraj + Kavya). Generates the static synthetic personas + benchmark JSON ONCE,
checked into the repo. Zero runtime generation during the demo (CONTRACT-005).

Run:  python scripts/generate_data.py
Out:  data/personas/*.json  (CONTRACT-001 shape, superset with a few seed-only fields)
      data/benchmark.json   (CONTRACT-005 shape)

Transaction convention (locked with TASK-003's burn-rate model):
  amount > 0  -> money IN  (category "opening_balance" or "income")
  amount < 0  -> money OUT (category in the 5 spend buckets below)
This keeps CONTRACT-003's `{ transactions }`-only input self-sufficient: a bank-statement style
ledger of signed credits/debits is everything the regressor needs to reconstruct running balance.
The classifier (CONTRACT-002) ignores income/opening_balance rows and reads only the 5 spend buckets.
"""

import json
import os
import random
from datetime import date, timedelta

SPEND_CATEGORIES = ["food", "shopping", "bills", "entertainment", "savings"]

# Human-readable descriptions per category, for a realistic-looking ledger.
DESCRIPTIONS = {
    "food": ["Swiggy order", "Zomato dinner", "Groceries — DMart", "Cafe Coffee Day",
             "Local restaurant", "BigBasket", "Street food", "Bakery"],
    "shopping": ["Myntra", "Amazon", "Flipkart", "Zara", "Decathlon", "Electronics store"],
    "bills": ["Electricity bill", "Mobile recharge", "Broadband", "Rent share",
              "Water bill", "Gas cylinder"],
    "entertainment": ["BookMyShow", "Netflix", "Spotify", "Pub night", "Concert ticket",
                      "Gaming top-up", "Weekend trip"],
    "savings": ["SIP transfer", "RD deposit", "Emergency fund", "Gold savings"],
}

# Per-persona spec: intended archetype, income, opening liquid balance, monthly spend as a
# fraction of income, and the category split (must sum to 1.0). Splits are chosen close to the
# TASK-003 centroids so classification is unambiguous but not identical (real data has noise).
PERSONAS = [
    {
        "session_id": "persona-priya",
        "persona": "Priya — the Disciplined Saver",
        "archetype_intended": "Disciplined Saver",
        "monthly_income": 70000,
        "opening_balance": 120000,
        "spend_fraction": 0.60,      # spends 60% of income, banks the rest -> balance grows
        "income_bracket": "50k-75k",
        "city_tier": "Tier 1",
        "split": {"food": 0.16, "shopping": 0.10, "bills": 0.30, "entertainment": 0.06, "savings": 0.38},
        "seed": 11,
    },
    {
        "session_id": "persona-rahul",
        "persona": "Rahul — the Foodie",
        "archetype_intended": "The Foodie",
        "monthly_income": 55000,
        "opening_balance": 45000,
        "spend_fraction": 0.92,      # spends most of income -> modest runway
        "income_bracket": "50k-75k",
        "city_tier": "Tier 1",
        "split": {"food": 0.42, "shopping": 0.14, "bills": 0.20, "entertainment": 0.15, "savings": 0.09},
        "seed": 22,
    },
    {
        "session_id": "persona-aisha",
        "persona": "Aisha — the Impulsive Spender",
        "archetype_intended": "Impulsive Spender",
        "monthly_income": 60000,
        "opening_balance": 30000,
        "spend_fraction": 1.12,      # overspends -> balance declines, short runway
        "income_bracket": "50k-75k",
        "city_tier": "Tier 2",
        "split": {"food": 0.19, "shopping": 0.37, "bills": 0.19, "entertainment": 0.20, "savings": 0.05},
        "seed": 33,
    },
    {
        "session_id": "persona-riya",
        "persona": "Riya — the Social Butterfly",
        "archetype_intended": "Social Butterfly",
        "monthly_income": 68000,
        "opening_balance": 55000,
        "spend_fraction": 0.88,
        "income_bracket": "50k-75k",
        "city_tier": "Tier 2",
        "split": {"food": 0.18, "shopping": 0.14, "bills": 0.22, "entertainment": 0.38, "savings": 0.08},
        "seed": 44,
    },
    {
        "session_id": "persona-arjun",
        "persona": "Arjun — the Balanced Spender",
        "archetype_intended": "Balanced Spender",
        "monthly_income": 85000,
        "opening_balance": 90000,
        "spend_fraction": 0.78,
        "income_bracket": "75k-100k",
        "city_tier": "Tier 1",
        "split": {"food": 0.21, "shopping": 0.16, "bills": 0.27, "entertainment": 0.13, "savings": 0.23},
        "seed": 55,
    },
]

# Roughly how many transactions per month per category (frequency patterns differ by archetype).
FREQ = {"food": 18, "shopping": 6, "bills": 5, "entertainment": 8, "savings": 2}

MONTHS = 3
START = date(2026, 4, 1)  # 3-month window ending ~2026-06-30


def _month_starts(start: date, months: int):
    out = []
    y, m = start.year, start.month
    for _ in range(months):
        out.append(date(y, m, 1))
        m += 1
        if m > 12:
            m = 1
            y += 1
    return out


def _split_amount(total: float, n: int, rng: random.Random) -> list:
    """Split `total` into n positive chunks with some variance (never zero)."""
    if n <= 0:
        return []
    weights = [rng.uniform(0.6, 1.4) for _ in range(n)]
    s = sum(weights)
    return [round(total * w / s) for w in weights]


def build_persona(spec: dict) -> dict:
    rng = random.Random(spec["seed"])
    monthly_income = spec["monthly_income"]
    monthly_spend = monthly_income * spec["spend_fraction"]
    txns = []

    # Opening balance as the first ledger row.
    txns.append({
        "date": (START - timedelta(days=1)).isoformat(),
        "category": "opening_balance",
        "amount": spec["opening_balance"],
        "description": "Opening balance",
    })

    for ms in _month_starts(START, MONTHS):
        # Salary credit on the 1st of each month.
        txns.append({
            "date": ms.isoformat(),
            "category": "income",
            "amount": monthly_income,
            "description": "Monthly salary",
        })
        # Expenses per category, spread across the month.
        for cat in SPEND_CATEGORIES:
            cat_total = monthly_spend * spec["split"][cat]
            n = max(1, FREQ[cat] + rng.randint(-1, 1))
            for amt in _split_amount(cat_total, n, rng):
                if amt <= 0:
                    continue
                day = rng.randint(1, 27)
                txns.append({
                    "date": date(ms.year, ms.month, day).isoformat(),
                    "category": cat,
                    "amount": -amt,  # money out
                    "description": rng.choice(DESCRIPTIONS[cat]),
                })

    txns.sort(key=lambda t: t["date"])

    return {
        "session_id": spec["session_id"],
        "persona": spec["persona"],
        "monthly_income": monthly_income,
        "income_bracket": spec["income_bracket"],
        "city_tier": spec["city_tier"],
        "archetype_intended": spec["archetype_intended"],  # sanity-check reference, not a contract field
        "archetype": None,
        "burn_rate": None,
        "transactions": txns,
        "goal": None,
        "last_decide_verdict": None,
        "benchmark": None,
    }


# ---- CONTRACT-005 benchmark: 5 income brackets x 3 city tiers, no gaps ----
INCOME_BRACKETS = ["0-25k", "25k-50k", "50k-75k", "75k-100k", "100k+"]
CITY_TIERS = ["Tier 1", "Tier 2", "Tier 3"]
BENCH_CATEGORIES = ["food", "shopping", "bills", "entertainment", "savings"]

# Base monthly INR percentiles per category (Tier 1, mid bracket). Scaled per bracket/tier below.
BASE_PERCENTILES = {
    "food":          {"p25": 4000, "p50": 6500, "p75": 9000, "p90": 13000},
    "shopping":      {"p25": 2000, "p50": 4000, "p75": 7000, "p90": 12000},
    "bills":         {"p25": 5000, "p50": 8000, "p75": 12000, "p90": 18000},
    "entertainment": {"p25": 1500, "p50": 3000, "p75": 5500, "p90": 9000},
    "savings":       {"p25": 3000, "p50": 8000, "p75": 15000, "p90": 25000},
}
BRACKET_SCALE = {"0-25k": 0.5, "25k-50k": 0.75, "50k-75k": 1.0, "75k-100k": 1.35, "100k+": 1.8}
TIER_SCALE = {"Tier 1": 1.0, "Tier 2": 0.8, "Tier 3": 0.62}


def build_benchmark() -> list:
    rows = []
    for bracket in INCOME_BRACKETS:
        for tier in CITY_TIERS:
            scale = BRACKET_SCALE[bracket] * TIER_SCALE[tier]
            cats = []
            for cat in BENCH_CATEGORIES:
                base = BASE_PERCENTILES[cat]
                cats.append({
                    "category": cat,
                    "percentiles": {k: round(v * scale) for k, v in base.items()},
                })
            rows.append({"income_bracket": bracket, "city_tier": tier, "categories": cats})
    return rows


def main():
    root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    personas_dir = os.path.join(root, "data", "personas")
    os.makedirs(personas_dir, exist_ok=True)

    for spec in PERSONAS:
        p = build_persona(spec)
        path = os.path.join(personas_dir, f"{spec['session_id']}.json")
        with open(path, "w") as f:
            json.dump(p, f, indent=2)
        n_spend = sum(1 for t in p["transactions"] if t["category"] in SPEND_CATEGORIES)
        print(f"{spec['session_id']}: {len(p['transactions'])} txns ({n_spend} spend) -> {path}")

    bench = {
        "_note": "Static peer-benchmark data (CONTRACT-005). Generated once by scripts/generate_data.py, "
                 "never at runtime. 5 income brackets x 3 city tiers, no gaps.",
        "brackets": build_benchmark(),
    }
    bench_path = os.path.join(root, "data", "benchmark.json")
    with open(bench_path, "w") as f:
        json.dump(bench, f, indent=2)
    print(f"benchmark: {len(bench['brackets'])} bracket/tier rows -> {bench_path}")


if __name__ == "__main__":
    main()

File: scripts/generate-data.mjs
// FORE — scripts/generate-data.mjs
// Generates the static synthetic data (data/personas/*.json + data/benchmark.json) ONCE.
// Deterministic (seeded PRNG) so re-running produces identical output. The JSON is checked in;
// nothing is ever generated at runtime during the demo (TASK-004 constraint / CONTRACT-005).
//
// Run: node scripts/generate-data.mjs

import { writeFileSync, mkdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");

// ---------- deterministic PRNG (mulberry32) ----------
function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// ---------- date helpers (90-day window ending just before the demo day) ----------
const WINDOW_END = new Date("2026-07-15T00:00:00Z");
const WINDOW_DAYS = 90;

function isoDay(offsetFromStart) {
  const d = new Date(WINDOW_END);
  d.setUTCDate(d.getUTCDate() - (WINDOW_DAYS - 1 - offsetFromStart));
  return d.toISOString().slice(0, 10);
}

// ---------- persona generation ----------
// Three visibly distinct spending patterns (TASK-004: "not uniform random" — distinctness is what
// makes TASK-003's classification meaningful).

function genPersona({ seed, persona, displayName, monthlyIncome, incomeBracket, cityTier, plan }) {
  const rand = mulberry32(seed);
  const between = (lo, hi) => lo + rand() * (hi - lo);
  const roundTo = (v, step) => Math.round(v / step) * step;
  const transactions = [];

  for (let day = 0; day < WINDOW_DAYS; day++) {
    const date = isoDay(day);
    const dayOfMonth = Number(date.slice(8, 10));

    // Fixed monthly bills on set days of the month.
    for (const bill of plan.monthly) {
      if (dayOfMonth === bill.day) {
        transactions.push({
          date,
          category: bill.category,
          amount: roundTo(between(bill.amount * 0.97, bill.amount * 1.03), 10),
          description: bill.description,
        });
      }
    }

    // Probabilistic daily spends per category.
    for (const c of plan.daily) {
      if (rand() < c.chance) {
        transactions.push({
          date,
          category: c.category,
          amount: roundTo(between(c.min, c.max), 5),
          description: c.descriptions[Math.floor(rand() * c.descriptions.length)],
        });
        // Occasional same-day double spend for "spiky" personas.
        if (c.doubleChance && rand() < c.doubleChance) {
          transactions.push({
            date,
            category: c.category,
            amount: roundTo(between(c.min, c.max), 5),
            description: c.descriptions[Math.floor(rand() * c.descriptions.length)],
          });
        }
      }
    }
  }

  transactions.sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : 0));
  return {
    persona,
    display_name: displayName,
    monthly_income: monthlyIncome,
    income_bracket: incomeBracket,
    city_tier: cityTier,
    transactions,
  };
}

const personas = [
  genPersona({
    seed: 20260415,
    persona: "arjun",
    displayName: "Arjun — steady salaried engineer",
    monthlyIncome: 90000,
    incomeBracket: "60k-100k",
    cityTier: "tier-1",
    plan: {
      monthly: [
        { day: 1, category: "rent", amount: 22000, description: "Monthly rent" },
        { day: 5, category: "utilities", amount: 2600, description: "Electricity + internet" },
      ],
      daily: [
        { category: "groceries", chance: 0.42, min: 450, max: 950, descriptions: ["BigBasket order", "Local kirana", "DMart run"] },
        { category: "transport", chance: 0.62, min: 80, max: 220, descriptions: ["Metro card top-up", "Auto fare", "Office cab share"] },
        { category: "food_dining", chance: 0.14, min: 350, max: 750, descriptions: ["Weekend dinner out", "Team lunch"] },
        { category: "entertainment", chance: 0.05, min: 250, max: 600, descriptions: ["Movie ticket", "OTT subscription"] },
        { category: "shopping", chance: 0.04, min: 600, max: 1800, descriptions: ["Clothing", "Household item"] },
      ],
    },
  }),
  genPersona({
    seed: 20260416,
    persona: "meera",
    displayName: "Meera — the foodie designer",
    monthlyIncome: 55000,
    incomeBracket: "30k-60k",
    cityTier: "tier-2",
    plan: {
      monthly: [
        { day: 2, category: "rent", amount: 13000, description: "Monthly rent" },
        { day: 6, category: "utilities", amount: 1800, description: "Electricity + internet" },
      ],
      daily: [
        { category: "food_dining", chance: 0.85, min: 180, max: 720, doubleChance: 0.3, descriptions: ["Swiggy order", "Cafe brunch", "Street food crawl", "Zomato dinner", "New restaurant try"] },
        { category: "groceries", chance: 0.16, min: 300, max: 700, descriptions: ["Local kirana", "Veggie market"] },
        { category: "transport", chance: 0.32, min: 60, max: 180, descriptions: ["Auto fare", "Bus pass", "Rapido"] },
        { category: "entertainment", chance: 0.07, min: 200, max: 500, descriptions: ["Movie ticket", "Food festival entry"] },
        { category: "shopping", chance: 0.05, min: 400, max: 1500, descriptions: ["Kitchen gadget", "Cookbook"] },
      ],
    },
  }),
  genPersona({
    seed: 20260417,
    persona: "rohan",
    displayName: "Rohan — impulsive gadget lover",
    monthlyIncome: 130000,
    incomeBracket: "100k-200k",
    cityTier: "tier-1",
    plan: {
      monthly: [
        { day: 1, category: "rent", amount: 30000, description: "Monthly rent" },
        { day: 4, category: "utilities", amount: 3200, description: "Electricity + internet" },
      ],
      daily: [
        { category: "shopping", chance: 0.3, min: 500, max: 9000, doubleChance: 0.15, descriptions: ["Amazon impulse buy", "Flipkart sale grab", "Sneaker drop", "Gadget accessory", "Flash-sale checkout"] },
        { category: "entertainment", chance: 0.26, min: 300, max: 1400, descriptions: ["Concert ticket", "Gaming credits", "Club night", "Go-karting"] },
        { category: "food_dining", chance: 0.3, min: 400, max: 1200, descriptions: ["Zomato order", "Brewery night", "Fine dining"] },
        { category: "groceries", chance: 0.12, min: 400, max: 900, descriptions: ["Instamart order", "DMart run"] },
        { category: "transport", chance: 0.36, min: 120, max: 450, descriptions: ["Uber ride", "Fuel top-up"] },
      ],
    },
  }),
];

// ---------- benchmark JSON (CONTRACT-005) ----------
// All 5 income brackets x 3 city tiers, no gaps (TASK-004 edge case). Each element matches the
// CONTRACT-005 object shape exactly; the file is an array covering every combination.

const BRACKETS = [
  { id: "0-30k", mid: 20000 },
  { id: "30k-60k", mid: 45000 },
  { id: "60k-100k", mid: 80000 },
  { id: "100k-200k", mid: 150000 },
  { id: "200k+", mid: 260000 },
];
const TIERS = [
  { id: "tier-1", mult: 1.25 },
  { id: "tier-2", mult: 1.0 },
  { id: "tier-3", mult: 0.8 },
];
// Share of mid-income spent per category at the median, per bracket (spend share falls as income rises).
const CATEGORY_SHARE = {
  food_dining: 0.085,
  groceries: 0.09,
  transport: 0.05,
  entertainment: 0.04,
  shopping: 0.065,
  utilities: 0.045,
};

const benchRand = mulberry32(20260701);
const benchmark = [];
for (const b of BRACKETS) {
  for (const t of TIERS) {
    const categories = Object.entries(CATEGORY_SHARE).map(([category, share]) => {
      const jitter = 0.9 + benchRand() * 0.2;
      const p50 = Math.round((b.mid * share * t.mult * jitter) / 50) * 50;
      return {
        category,
        percentiles: {
          p25: Math.round((p50 * 0.62) / 50) * 50,
          p50,
          p75: Math.round((p50 * 1.45) / 50) * 50,
          p90: Math.round((p50 * 2.1) / 50) * 50,
        },
      };
    });
    benchmark.push({ income_bracket: b.id, city_tier: t.id, categories });
  }
}

// ---------- write files ----------
mkdirSync(join(ROOT, "data", "personas"), { recursive: true });
for (const p of personas) {
  writeFileSync(join(ROOT, "data", "personas", `${p.persona}.json`), JSON.stringify(p, null, 2) + "\n");
  console.log(`data/personas/${p.persona}.json — ${p.transactions.length} transactions`);
}
writeFileSync(join(ROOT, "data", "benchmark.json"), JSON.stringify(benchmark, null, 2) + "\n");
console.log(`data/benchmark.json — ${benchmark.length} bracket/tier combinations`);
